### PR TITLE
feat(front/back): Reset password flow

### LINKED ISSUE #19: Reset password flow
**frontend**
- the user clicks on the `forgot password` link
- the user is redirected to `/forgot`
- the user enters his email and clicks on `send password reset email`
- the mutation `forgotPassword(email: String!)` is called
- display `Email sent!`

**backend**
- generate token
- save token in the database
- send an email with the reset link (`https://c3pm.io/reset?token=xxx`)

**frontend**
- the user clicks on the link in the email
- the user enters a new password and confirms
- the user clicks on `reset my password`
- the mutation `resetPassword(token: String!, password: String!)` is called

**backend**
- hash le password
- update hashed password where `token=token`

**frontend**
- the user is redirected to `/login`

### PR DESCRIPTION
**Pull request by angela-hubner**

### CODE REVIEW

File: back/src/services/ForgotService.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
+import { v4 as uuidv4 } from 'uuid';
+import { User } from '@prisma/client';
+import sgMail = require('@sendgrid/mail');
```
Comment (gabrielcolson) [inline back/src/services/ForgotService.ts]:
```suggestion
import * as sgMail from "@sendgrid/mail"
```

**Review by chloebourbion — CHANGES_REQUESTED** [review]
image
On mobile can you make the input and the button take all the width and make the text smaller?

What do you think of "An email has been send! check your inbox" as an explanation sentence?

File: front/pages/forgot_password.tsx [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    justifyContent: 'center',
+    alignItems: 'center',
+    [theme.breakpoints.down('sm')]: {
+      paddingTop: '8vw',
+    },
+  },
+  form: {
+    display: 'flex',
+    width: '20%',
+    margin: 'auto',
```
Comment (chloebourbion) [inline front/pages/forgot_password.tsx]: you shouldn't use `margin: auto` to center your content, try `alignItems: 'center'` with a `width: 100%` and apply a width to your input and button if you don't want them to take all the page (or maybe removing the `fullWidth` props is enough). You can also wrap your input and button in a `div` if you don't want to/cannot apply a width to the input/button easily

File: front/pages/forgot_password.tsx [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+import { ForgotParams, forgotSchema } from 'utils/validation';
+
+const useStyles = makeStyles((theme) => ({
+  text: {
+    display: 'flex',
+    justifyContent: 'center',
+    marginTop: theme.spacing(3),
+    marginBottom: theme.spacing(5),
+    minWidth: 175,
+    fontSize: 24,
```
Comment (chloebourbion) [inline front/pages/forgot_password.tsx]: can you reduce the fontSize just a little?

**Review by chloebourbion — CHANGES_REQUESTED** [review]

File: back/src/graphql/types/User.ts [hunk 4]
```diff
@@ -7,6 +7,7 @@ export const User = objectType({
     t.model.username();
     t.model.email();
     t.model.description();
+    t.model.resetPasswordToken();
```
Comment (gabrielcolson) [inline back/src/graphql/types/User.ts]: remove this

File: back/.envrc.example [hunk 5]
```diff
@@ -6,3 +6,4 @@ export SESSION_SECRET=toto
 export REGISTRY_HOST=localhost
 export REGISTRY_PORT=4444
 export REGISTRY_SECRET=highfive
+export FORGOT_URL=http://localhost:3000/reset_password?token=
```
Comment (gabrielcolson) [inline back/.envrc.example]: change this to `FRONTEND_URL` and use only the base URL

File: back/src/services/ForgotService.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+      html: `<p>Click <a href='${process.env.FORGOT_URL}${token}'>here</a> to reset your password.</p>`,
+    });
+    return this.db.user.update({
+      where: {
+        email,
+      },
+      data: {
+        resetPasswordToken: token,
+      },
+    });
```
Comment (gabrielcolson) [inline back/src/services/ForgotService.ts]: Switch both statements and check that the user exists before sending him an email

File: front/package.json [hunk 7]
```diff
@@ -41,6 +41,7 @@
     "react-markdown": "^4.3.1",
     "react-scroll": "^1.8.1",
     "semver": "^7.3.2",
+    "uuid": "^8.3.1",
```
Comment (gabrielcolson) [inline front/package.json]: Remove this unused dependency

File: front/pages/forgot_password.tsx [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+            >
+              Enter your email and click the button below to receive a link to reset your password.
+            </Typography>
+          )
+      }
+      <Formik
+        initialValues={{ email: '' }}
+        validationSchema={forgotSchema}
+        onSubmit={(email: ForgotParams): void => {
+          forgot(email);
```
Comment (gabrielcolson) [inline front/pages/forgot_password.tsx]: missing await

File: front/pages/reset_password.tsx [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+        variant="subtitle1"
+        className={classes.text}
+      >
+        Enter your new password and password confirmation.
+      </Typography>
+      <Formik
+        initialValues={{ password: '', confirm: '' }}
+        validationSchema={resetSchema}
+        onSubmit={async (params: ResetParams): Promise<void> => {
+          reset({ token: token.toString(), password: params.password });
```
Comment (gabrielcolson) [inline front/pages/reset_password.tsx]: missing await

**Review by gabrielcolson — CHANGES_REQUESTED** [review]

**Review by gabrielcolson — APPROVED** [review]

**Review by chloebourbion — APPROVED** [review]

**Pull request merged**